### PR TITLE
Optimized Paged Attention for HPU

### PR DESCRIPTION
**Pull request by kzawora-intel**
This PR adds Paged Attention implementation for Gaudi2. In standalone benchmark (B=32, Mkv=1024, Hkv=32, K=128) it showed 200x performance improvement over current implementation:

Current (naive) version:
```
INFO:root:[REQ:0][B:32, Mq:1, Mkv:996, Hq:32, Hkv:32, K:128] paged_attention_op_time (hpu): 106.500s
INFO:root:[REQ:1][B:32, Mq:1, Mkv:1006, Hq:32, Hkv:32, K:128] paged_attention_op_time (hpu): 46.476s
INFO:root:[REQ:2][B:32, Mq:1, Mkv:1015, Hq:32, Hkv:32, K:128] paged_attention_op_time (hpu): 21.430s
INFO:root:[REQ:3][B:32, Mq:1, Mkv:953, Hq:32, Hkv:32, K:128] paged_attention_op_time (hpu): 12.854s
INFO:root:[REQ:4][B:32, Mq:1, Mkv:1008, Hq:32, Hkv:32, K:128] paged_attention_op_time (hpu): 9.996s
INFO:root:[REQ:5][B:32, Mq:1, Mkv:980, Hq:32, Hkv:32, K:128] paged_attention_op_time (hpu): 5.928s
INFO:root:[REQ:6][B:32, Mq:1, Mkv:1024, Hq:32, Hkv:32, K:128] paged_attention_op_time (hpu): 8.595s
INFO:root:[REQ:7][B:32, Mq:1, Mkv:1015, Hq:32, Hkv:32, K:128] paged_attention_op_time (hpu): 5.835s
INFO:root:[REQ:8][B:32, Mq:1, Mkv:995, Hq:32, Hkv:32, K:128] paged_attention_op_time (hpu): 6.296s
INFO:root:[REQ:9][B:32, Mq:1, Mkv:1023, Hq:32, Hkv:32, K:128] paged_attention_op_time (hpu): 5.833s
INFO:root:[ALL:10][B:(32-32), Mq:1, Mkv:(953-1024), Hq:32, Hkv:32, K:128] paged_attention_op_time (hpu): 229.758s
```

Optimized version:
```
INFO:root:[REQ:0][B:32, Mq:1, Mkv:996, Hq:32, Hkv:32, K:128] paged_attention_op_time (hpu): 0.959s
INFO:root:[REQ:1][B:32, Mq:1, Mkv:1006, Hq:32, Hkv:32, K:128] paged_attention_op_time (hpu): 0.022s
INFO:root:[REQ:2][B:32, Mq:1, Mkv:1015, Hq:32, Hkv:32, K:128] paged_attention_op_time (hpu): 0.020s
INFO:root:[REQ:3][B:32, Mq:1, Mkv:953, Hq:32, Hkv:32, K:128] paged_attention_op_time (hpu): 0.020s
INFO:root:[REQ:4][B:32, Mq:1, Mkv:1008, Hq:32, Hkv:32, K:128] paged_attention_op_time (hpu): 0.020s
INFO:root:[REQ:5][B:32, Mq:1, Mkv:980, Hq:32, Hkv:32, K:128] paged_attention_op_time (hpu): 0.020s
INFO:root:[REQ:6][B:32, Mq:1, Mkv:1024, Hq:32, Hkv:32, K:128] paged_attention_op_time (hpu): 0.020s
INFO:root:[REQ:7][B:32, Mq:1, Mkv:1015, Hq:32, Hkv:32, K:128] paged_attention_op_time (hpu): 0.020s
INFO:root:[REQ:8][B:32, Mq:1, Mkv:995, Hq:32, Hkv:32, K:128] paged_attention_op_time (hpu): 0.021s
INFO:root:[REQ:9][B:32, Mq:1, Mkv:1023, Hq:32, Hkv:32, K:128] paged_attention_op_time (hpu): 0.020s
INFO:root:[ALL:10][B:(32-32), Mq:1, Mkv:(953-1024), Hq:32, Hkv:32, K:128] paged_attention_op_time (hpu): 1.149s
```

### CODE REVIEW

File: vllm/hpu/ops.py [hunk 1]
```diff
@@ -25,72 +25,247 @@ def gelu_new(output, input):
...
+    seq_context_len = torch.tensor([0], dtype=torch.int64, device=device)
+    for i in range(0, max_num_blocks_per_seq):  # can run in parallel
+        with torch.profiler.record_function(f"block_loop"):
+            with torch.profiler.record_function("seq_index_fill"):
+                seq_index.fill_(0)
+
+            ## hard override for filler. These blocks would contribute nothing to the output due to zero attention_probs and will clog up compute resources
+            # with torch.profiler.record_function('block_seq_len_check'):
+            #    if (block_index - 2) * block_size > torch.max(context_lens):
+            #        break
```
Comment (mdvoretc-intel) [inline vllm/hpu/ops.py:169]: Please remove this block. It's unlikely to come back into use, even if a new check against overcomputing is introduced.

File: vllm/hpu/ops.py [hunk 2]
```diff
@@ -25,72 +25,247 @@ def gelu_new(output, input):
...
+        (max_num_blocks_per_seq, num_seqs, num_query_heads, 1, block_size), torch.finfo(query.dtype).min, dtype=query.dtype, device=device
+    )
+    value_blocks = torch.zeros((max_num_blocks_per_seq, num_seqs, num_kv_heads, block_size, head_size), dtype=query.dtype, device=device)
+    seq_index = torch.tensor([0], dtype=torch.int64, device=device)
+    block_index = torch.tensor([0], dtype=torch.int64, device=device)
+    seq_block_table_idx = torch.tensor([0], dtype=torch.int64, device=device)
+    seq_context_len = torch.tensor([0], dtype=torch.int64, device=device)
+    for i in range(0, max_num_blocks_per_seq):  # can run in parallel
+        with torch.profiler.record_function(f"block_loop"):
+            with torch.profiler.record_function("seq_index_fill"):
```
Comment (mdvoretc-intel) [inline vllm/hpu/ops.py:163]: Do these profiler labels have any effect when not profiling this code?

File: vllm/hpu/ops.py [hunk 3]
```diff
@@ -25,72 +25,247 @@ def gelu_new(output, input):
...
+            ## hard override for filler. These blocks would contribute nothing to the output due to zero attention_probs and will clog up compute resources
+            # with torch.profiler.record_function('block_seq_len_check'):
+            #    if (block_index - 2) * block_size > torch.max(context_lens):
+            #        break
+
+            with torch.profiler.record_function("slice_weights_values"):
+                # single block attn weight of shape [B, Hq, Mq(=1), block_size], equivalent to attn_weights_blocks[i]
+                attn_weights = attn_weights_blocks.index_select(0, block_index).squeeze(0)
+
+            with torch.profiler.record_function("fetch_block_table"):
```
Comment (mdvoretc-intel) [inline vllm/hpu/ops.py:175]: The prevalence of the profiler labels is reducing code readability, especially in this part of the code where each `record_function` covers a single statement.

File: vllm/hpu/ops.py [hunk 4]
```diff
@@ -25,72 +25,247 @@ def gelu_new(output, input):
...
+    block_size,
+    max_context_len,
+    alibi_slopes,
+    attn_masks=None,
+) -> None:
+    sanitize_values = True
+    device = query_in.device
+    query = query_in
+    key_cache = key_cache_in
+    value_cache = value_cache_in
```
Comment (mdvoretc-intel) [inline vllm/hpu/ops.py:141]: The `*_in` args were used in a prior version to facilitate type casting, since FP16 was causing precision issues in softmax. If that is no longer the case, then the arg names should be changed and these lines should be removed.

File: vllm/hpu/ops.py [hunk 5]
```diff
@@ -25,72 +25,247 @@ def gelu_new(output, input):
...
+            with torch.profiler.record_function("store_block_weight"):
+                attn_weights_blocks.index_copy_(0, block_index, attn_weights.unsqueeze(0))
+
+        with torch.profiler.record_function("block_index_inc"):
+            block_index.add_(1)
+        if device == "hpu":
+            htorch.core.mark_step()
+
+    # Cleanup out-of-bound weights and values
+    attn_weights_blocks_filler = torch.finfo(query.dtype).min
```
Comment (mdvoretc-intel) [inline vllm/hpu/ops.py:232]: The "minimum value of query type" filler is also used elsewhere in this code (e.g. the `attn_weights_blocks` definition) and should probably be defined and commented on at the start of the function. The question also arises as to the type restrictions, since the required behavior for this filler is to produce a zero under `exp()`, and that behavior is not observed on FP16.

File: vllm/hpu/ops.py [hunk 6]
```diff
@@ -25,72 +25,247 @@ def gelu_new(output, input):
...
+    scale,
+    block_tables,
+    context_lens,
+    block_size,
+    max_context_len,
+    alibi_slopes,
+    attn_masks=None,
+) -> None:
+    sanitize_values = True
+    device = query_in.device
```
Comment (mdvoretc-intel) [inline vllm/hpu/ops.py:138]: Does this statement produce the device type ("hpu") or a device identifier ("hpu:0")? This is relevant to its use in further checks.

File: vllm/hpu/ops.py [hunk 7]
```diff
@@ -25,72 +25,247 @@ def gelu_new(output, input):
...
+            with torch.profiler.record_function("reshape_for_gemm"):
+                attention_probs = attention_probs.to(value.dtype).reshape(num_seqs * num_query_heads, 1, block_size)
+                value = value.reshape(num_seqs * num_query_heads, block_size, head_size)
+            with torch.profiler.record_function("gemm_out"):
+                out = torch.matmul(attention_probs, value).reshape(num_seqs, num_query_heads, head_size)
+            with torch.profiler.record_function("gemm_accumulate"):
+                output.add_(out)
+    if device == "hpu":
+        htorch.core.mark_step()
     return output.to(dtype=query_in.dtype)
```
Comment (mdvoretc-intel) [inline vllm/hpu/ops.py:269]: There's no type casting of the query in the current code, so this casting of the output is superfluous.

**Comment by madamczykhabana** [comment]
Since #14 was merged, I guess we can close this.
